packages/upset/src/provenance/index.ts

```diff
@@ -19,108 +21,102 @@ const firstAggAction = createAction<UpsetConfig, [AggregateBy], Events>(
   },
 );
 
-const firstOverlapAction = createAction<UpsetConfig, [number], Events>(
+const firstOverlapAction = registry.register('first-overlap',
   (state, overlap) => {
     state.firstOverlapDegree = overlap;
     return state;
   },
 );
 
-const secondAggAction = createAction<UpsetConfig, [AggregateBy], Events>(
+const secondAggAction = registry.register('second-agg',
   (state, aggBy) => {
     state.secondAggregateBy = aggBy;
     return state;
   },
 );
 
-const secondOverlapAction = createAction<UpsetConfig, [number], Events>(
+const secondOverlapAction = registry.register('second-overlap',
   (state, overlap) => {
     state.secondOverlapDegree = overlap;
     return state;
   },
 );
 
-const sortByAction = createAction<UpsetConfig, [SortBy], Events>(
+const sortByAction = registry.register('sort-by',
   (state, sort) => {
     state.sortBy = sort;
     return state;
   },
 );
 
-const maxVisibleAction = createAction<UpsetConfig, [number], Events>(
+const maxVisibleAction = registry.register('max-visible',
   (state, maxVisible) => {
     state.filters.maxVisible = maxVisible;
     return state;
   },
 );
 
-const minVisibleAction = createAction<UpsetConfig, [number], Events>(
+const minVisibleAction = registry.register('min-visible',
   (state, minVisible) => {
     state.filters.minVisible = minVisible;
     return state;
   },
 );
 
-const hideEmptyAction = createAction<UpsetConfig, [boolean], Events>(
+const hideEmptyAction = registry.register('hide-empty',
   (state, hide) => {
     state.filters.hideEmpty = hide;
     return state;
   },
 );
 
-const addToVisibleAction = createAction<UpsetConfig, [string], Events>(
+const addToVisibleAction = registry.register('add-to-visible',
   (state, newSet) => {
     const newSets = new Set([...state.visibleSets, newSet]);
     state.visibleSets = Array.from(newSets);
     return state;
   },
 );
 
-const removeFromVisibleAction = createAction<UpsetConfig, [string], Events>(
+const removeFromVisibleAction = registry.register('remove-from-visible',
   (state, newSet) => {
-    state.visibleSets = state.visibleSets.filter((v) => v !== newSet);
+    state.visibleSets = state.visibleSets.filter((v: string) => v !== newSet);
```

Comment:
You might not need to specify string here if there are better types on `state.visibleSets`